Propagate CQL coordinator timeouts and failures for reads

piodul opened the pull request:
This PR propagates the read coordinator logic so that read timeout and read failure exceptions are propagated without throwing on the coordinator side.

This PR is only concerned with exceptions which were originally thrown by the coordinator (in read resolvers). Exceptions propagated through RPC and RPC timeouts will still throw, although those exceptions will be caught and converted into exceptions-as-values by read resolvers.

This is a continuation of work started in #10014.

Results of `perf_simple_query --smp 1 --operations-per-shard 1000000` (read workload), compared with merge base (10880fb0a720e3dcddcb301f8f1c624923e89a03):

```
BEFORE:
125085.13 tps ( 80.2 allocs/op,  12.2 tasks/op,   49010 insns/op)
125645.88 tps ( 80.2 allocs/op,  12.2 tasks/op,   49008 insns/op)
126148.85 tps ( 80.2 allocs/op,  12.2 tasks/op,   49005 insns/op)
126044.40 tps ( 80.2 allocs/op,  12.2 tasks/op,   49005 insns/op)
125799.75 tps ( 80.2 allocs/op,  12.2 tasks/op,   49003 insns/op)

AFTER:
127557.21 tps ( 80.2 allocs/op,  12.2 tasks/op,   49197 insns/op)
127835.98 tps ( 80.2 allocs/op,  12.2 tasks/op,   49198 insns/op)
127749.81 tps ( 80.2 allocs/op,  12.2 tasks/op,   49202 insns/op)
128941.17 tps ( 80.2 allocs/op,  12.2 tasks/op,   49192 insns/op)
129276.15 tps ( 80.2 allocs/op,  12.2 tasks/op,   49182 insns/op)
```

The PR does not introduce additional allocations on the read happy-path. The number of instructions used grows by about 200 insns/op. The increase in TPS is probably just a measurement error.

piodul commented on utils/exception_container.hh:
```diff
@@ -89,6 +100,27 @@ public:
+    //
+    // CAUTION: This method is _not_ noexcept if you clone a container
+    // which was created on another shard!
+    exception_container clone() const {
```
Here is a small problem: I initially wrote the exception_container to be move only, but it turned out that there is a place in the code where I need to set the same exception to two promises (`_cl_promise` and `_done_promise`). It would be really nice if a class used for writing exception handling logic wouldn't throw on copy, so I considered the following options:

1. (Blasphemous, I know) Use std::shared_ptr. The main problem with it is that it uses an atomic for reference counting. However, this cost won't be paid on the happy path, and it can't be worse than std::exception_ptr because the latter is also copyable and thread-safe. In case when an exception _is_ returned using the exception_container, if the container is only moved around and never copied, then I think we will pay only one atomic operation per request (when the container is destroyed). In case of an exception storm, it can be a lot, though.

   Funnily enough, switching to std::shared_ptr makes the happy path faster by about 100 insns/op, probably because the container type is smaller (only a pointer, foreign_ptr holds pointer + shard ID).

2. Convert std::unique_ptr to lw_shared_ptr. If the container is copied on the same shard, then we only increase the reference count and that is cheap. However, if the container comes from another shard, then we perform a full copy, which involves an allocation. This requires adding a `copy_local` to seastar which I'm using here, I'll send a patch if we agree this is a good direction. This is my current approach, but I'm not sure what to do if the allocation fails (currently, copying can throw). I see two options:

	- Make copying noexcept and abort on allocation failure. This sounds hardcore, but that's what seastar does when it can't allocate a task. The exception container doesn't need much memory, only to hold a std::variant with an exception, so maybe it's not a problem.
	- Return an empty exception container. Currently, we treat an empty container as if it had a "bad access" exception inside, so maybe we can give a special meaning to it. Unfortunately, empty containers can also be created when a failed result which was moved out is accidentally returned.
	- Add an inline bool flag to exception_container that it failed to be constructed properly. This might increase the size of the exception container and make it more costly to move (I'll have to check).

Now that I wrote this, I realized that there already is a similar problem in the exception_container's constructor, which allocates and can fail.

Reviewers, what do you think? Which sounds the best?

avikivity replied:
Since std::exception_ptr is morally equivalent to std::shared_ptr, we don't lose compared to status quo ante.

libstdc++ recently optimized the destructor of std::shared_ptr [1] for the case of non-shared data (which will be the common case in the uncommon path of having an exception), so I think it's fine. We'll take a hit until we upgrade, but without real sharing, the hit is very small. 

[1] https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=dbf8bd3c2f2cd2d27ca4f0fe379bd9490273c6d7

piodul replied:
While for writes the optimization you mentioned will make it work well, in case of read failures/timeouts the shared_ptr will have to be copied, see here: [link](https://github.com/scylladb/scylla/pull/10092/commits/2fde37b4c317efe3b6f822a901b25b9096a54f7e#diff-b973f4e30060a6f4beaf81347658f565fa895998c849d92e80f574a6255bad75L2858-R2862). So, for read errors, we will actually pay the price of atomics even with the optimization that you mentioned (although it will be 1 atomic decrement less).

What about the exception_container's constructor? It can fail because it allocates space to hold the exception. How should it behave in case of bad_alloc?

avikivity replied:
I think it's okay.

For bad_alloc, it can store an std::exception_ptr containing a bad_alloc. We lose the timeout report, but if we can't allocate space for an exception, the system is dead anyway.

piodul replied:
I swtiched to std::shared_ptr and copying exception containers now does not throw.

As the problem of the throwing exception_container constructor is a bit orthogonal to this whole PR, I'll handle it in a followup.

gleb-cloudius commented on utils/result.hh:
```diff
@@ -44,4 +44,15 @@ concept ExceptionContainerResult = bo::is_basic_result<R>::value && ExceptionCon
 concept ExceptionContainerResultFuture = seastar::is_future<F>::value && ExceptionContainerResult<typename F::value_type>;
 
+template<typename L, typename R>
+concept ResultCompatible =
```
In what way are they compatible if value types can be completely different.

piodul replied:
They are compatible in the sense that they can hold the same error type. In a single piece of code you may want to use results which hold different value types but the same error type (e.g. `coordinator_result<T>` for different T's). This is useful because if two results have the same error type, you are sure you can convert one failed result to another.

Example:

```c++
template<typename T> using result = bo::result<T, unfunny_joke_error>;

result<pen> get_pen();
result<apple> get_apple();
result<apple_pen> make_apple_pen(apple a, pen p);

result<apple_pen> get_apple_pen() {
    result<pen> pen = get_pen();
    if (!pen) {
        // Conversions go like this:
        // result<pen> -> bo::failure_type<unfunny_joke_error> -> result<apple_pen>
        return pen.as_faliure();
    }
    result<apple> apple = get_apple();
    if (!apple) {
        return apple.as_failure();
    }
    return make_apple_pen(pen, apple);
}

```

gleb-cloudius replied:
But for that types they are holding should be convertible as well. I do not think the concept checks it.

piodul replied:
By "ResultCompatible" here I meant compatibility of the error types and not necessarily of values. Maybe the name is just confusing and suggests too much? How about `ResultWithCompatibleError`?

If by "types they are holding" you meant the error type, then yes, the concept is too restrictive. You could convert a failed `bo::result<T, E1>` to `bo::result<U, E2>` if the error type E1 converts to E2, but the concept does not allow it at the moment. I can relax it, although there is no real benefit of doing it now because we are only using `error_container`s now and they do not support conversions between each other yet.

gleb-cloudius replied:
> By "ResultCompatible" here I meant compatibility of the error types and not necessarily of values. Maybe the name is just confusing and suggests too much? 

It probably does since Avi also suggested to change it below.

> If by "types they are holding" you meant the error type, then yes, the concept is too restrictive. You could convert a failed bo::result<T, E1> to bo::result<U, E2> if the error type E1 converts to E2, but the concept does not allow it at the moment. I can relax it, although there is no real benefit of doing it now because we are only using error_containers now and they do not support conversions between each other yet.

No I meant that `bo::is_basic_result<L>::value` is not convertable to `bo::is_basic_result<R>::value`. So in this case how it is useful to know that error types are the same. You example will not work, no?

piodul replied:
I believe the example will work (after my ninja-edit, at least).

The `bo::result<T, E>` is a bit similar to `std::variant<T, E>`. If you know that your `std::variant<T, E>` holds an `E`, you can take it out, construct a `std::variant<U, E>` and return it.

It's similar for `bo::result<T, E>` -> `bo::result<U, E>`. Notice that I'm doing the `!apple` and `!pen` checks so that I'm sure that the results are holding an error. The `as_failure()` method is a type sugar which allows for the conversion to a compatible result type.

gleb-cloudius commented on utils/result_loop.hh:
```diff
@@ -251,4 +252,52 @@ result_map_reduce(Iterator begin, Iterator end, Mapper&& mapper, Initial initial
+            }
+        } else {
+            return ([] (future_type f, StopCondition stop_cond, AsyncAction action) -> seastar::future<result_type> {
+                for (;;) {
```
Why do you have two for() loops? Is it an optimization for the case where an action is immediately ready? Does it happen often?

piodul replied:
Yes, it's an optimization precisely for that case. I don't know how often it happens, I just wanted to have a result-aware equivalent of the `seastar::do_until` which has the same allocation behavior, and not have to worry about allocation count increasing when I substitute `seastar::do_until` with `utils::result_do_until`.

gleb-cloudius commented on utils/result_loop.hh:
```diff
@@ -251,4 +252,52 @@ result_map_reduce(Iterator begin, Iterator end, Mapper&& mapper, Initial initial
+                    if (!res) {
+                        co_return res;
+                    }
+                    if (stop_cond()) {
```
Why did you handle exceptions from the function in the outer loop but not here?

piodul replied:
The outer loop isn't executing in a coroutine, but here it is, so I thought an exception thrown from here will be converted to an exceptional future? To be fair I didn't write a test for this case, so I'll do and will check.

gleb-cloudius replied:
Right, they will.

avikivity commented on utils/result.hh:
```diff
@@ -44,4 +44,15 @@ concept ExceptionContainerResult = bo::is_basic_result<R>::value && ExceptionCon
+
+// Creates a compatible result type which can hold value type T or the same exceptional container as R
+template<typename T, ExceptionContainerResult R>
+using compatible_result = bo::result<T, typename R::error_type, exception_container_throw_policy>;
```
In std::allocator terminology, this is a rebind - we rebind the allocator from one type to another, but retaining the allocation method.

So I suggest to call this rebind_result. The value of reusing terminology is low when the terminology reused is obscure, but non-zero.

Let's also make the ResultCompatible concept less vague. I don't have a good suggestion though.

piodul replied:
Done, changed `compatible_result` -> `rebind_result` and `ResultCompatible` -> `ResultRebindableTo`.

avikivity commented on utils/result_loop.hh:
```diff
@@ -251,4 +252,52 @@ result_map_reduce(Iterator begin, Iterator end, Mapper&& mapper, Initial initial
+    using result_type = typename future_type::value_type;
+    static_assert(std::is_void_v<typename result_type::value_type>,
+            "The result type of the action must be future<result<void>>");
+    for (;;) {
```
do_until() is obsoleted by coroutines (it's a while loop). How would a coroutine using results look like? Isn't it just a while loop with an extra check on the result?

The only advantage do_until() has over a coroutine is zero allocations  if we never defer, but we give that up by using a coroutine here. I suggest to drop this and open-code it in the caller. It risks the caller not checking for errors, but that's a common risk with result types.

avikivity replied:
Sorry, I see you carefully avoided the allocation in case everything was ready.

So it really depends on the callers. If they are eligible for coroutine conversion (using do_with or expected to block), then they should be using coroutines. Otherwise, this.

avikivity commented on utils/result_loop.hh:
```diff
@@ -300,4 +300,46 @@ result_do_until(StopCondition stop_cond, AsyncAction action) {
+}
+inline
+auto result_repeat(AsyncAction&& action) noexcept {
+    using future_type = seastar::futurize_t<std::invoke_result_t<AsyncAction>>;
```
Let's evaluate the callers and see if open-coding makes sense there.

avikivity commented on utils/result_try.hh:
```diff
@@ -129,6 +141,18 @@ public:
+
+    std::exception_ptr clone_inner() {
+        return _eptr;
+    }
```
Would `clone_inner() &&` make sense?

piodul replied:
Wouldn't that imply that the exception is moved out of the handle? The purpose of `clone_inner` is to get an instance of exception_container/exception_ptr without consuming the one in the handle.

avikivity replied:
Yes of course, if it's an rvalue you'd just use it.

piodul replied:
Hmm, but then it wouldn't technically do any cloning, so I think the name would be confusing. Could you suggest a better name?

avikivity replied:
I meant I agree that `clone_inner() &&` is useless.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3679,9 +3679,12 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
-            try {
-                f.get();
+        (void)data_resolver->done().then_wrapped([this, exec, data_resolver, cmd = std::move(cmd), cl, timeout] (future<result<>> f) {
+            (void)utils::result_try([&] () -> result<> {
```
Why the cast to void? I guess bo::result is [[nodiscard]]?

Every cast to void deserves a common explaining why it is justified.

avikivity replied:
(the [[nodiscard]] isn't the justification, but that all errors are handled is).

piodul replied:
Yes, bo::result is nodiscard. I'll add comments with explanations.

I also had an idea to change result_try so that it does not return a result if it always handles all exceptions, i.e. it has catch_dots and none of the catch clauses return exceptions, but it's definitely out of scope of this PR (also, not sure if it wouldn't be too confusing).

gleb-cloudius replied:
Here I am also not sure if `result_try` is a simplification. You check for `res` to be an error explicitly anyway.

piodul replied:
@gleb-cloudius Won't explicit checking cause duplication of the exception handler? Although the handler is very simple, it is still duplication and you will have to remember about both of them if you want to introduce changes.

To be clear, I'm interpreting your suggestion as something like this:

```c++
try {
    result<> res = f.get();
    if (!res) {
        // Handler no. 1
        _result_promise.set_value(std::move(res).error());
        return;
    }

    /* ... lots of code here ... */

} catch (...) {
    // Handler no. 2
    _result_promise.set_exception(std::current_exception());
}
```

gleb-cloudius replied:
My point is that you already do explicit checking and then you re-check once again in the `try` wrapper. The idea of the `try` wrappers was to hide the diff between an error result and an exception, but here you forced to deal with the result anyway. I do not think resolving `_result_promise` twice is a big issue.

piodul replied:
Although I have to check the result explicitly, I'm not doing any real error handling there and just return it so that `result_try` handles it. The idea of the `result_try` helpers was not only to make commits easier to read, but also to handle both thrown exceptions and failed results in a unified way - you only need to write the `catch` handler once and it will work for both ways of reporting exceptions.

I won't insist on it too much - I can change error handling here in the way you wish - but I think there is a benefit to keeping the error logic defined in one place.

piodul replied:
Done, converted this code so that the `_result_promise` is set in two places.

piodul replied:
Added a comment for the void cast, too.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3700,18 +3699,20 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
-                        } catch (mutation_write_timeout_exception&) {
+                        return make_ready_future<::result<>>(bo::success());
+                    })).then_wrapped([this, exec] (future<::result<>>&& f) {
+                        (void)utils::result_try([&] {
```
(void) -> comment

piodul replied:
Added

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3784,12 +3784,16 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
+        (void)digest_resolver->has_cl().then_wrapped([exec, digest_resolver, timeout] (future<result<digest_read_result>> f) mutable {
             bool background_repair_check = false;
-            try {
+            (void)utils::result_try([&] () -> result<> {
```
(void) comment

gleb-cloudius replied:
Here as well. Is there an advantage to use `result_try`?

piodul replied:
I tried to get rid of `result_try` here, but the problem is that there is more code to execute after the try..catch block. After I handle the failed result, I need to jump right after the try..catch and continue. To do this, I could use a goto or an immediately invoked lambda - however I'm not sure if those options are really better than `result_try`, so if there are no objections I'll keep the `result_try` as it is here now.

piodul replied:
Added a (void) comment, too.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4070,19 +4070,25 @@ storage_proxy::query_result_local(schema_ptr s, lw_shared_ptr<query::read_comman
-        std::rethrow_exception(eptr);
-    } catch (read_timeout_exception& ex) {
+void storage_proxy::handle_read_error(std::variant<exceptions::coordinator_exception_container, std::exception_ptr> failure, bool range) {
+    (void)utils::result_try([&] () -> result<> {
```
void comment

piodul replied:
Added

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4134,7 +4134,12 @@ storage_proxy::query_singular(lw_shared_ptr<query::read_command> cmd,
+    // The following try..catch chain could be converted to an equivalent
+    // utils::result_futurize_try, however the code would no longer be placed
+    // inside a single coroutine and the number of task allocations would
+    // increase (utils::result_futurize_try is not a coroutine).
```
Indeed, this path needs all the speed it can get.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4148,25 +4153,35 @@ storage_proxy::query_singular(lw_shared_ptr<query::read_command> cmd,
+            result = co_await exec[0].first->execute(timeout);
+            if (result) {
+                handle_completion(exec[0]);
+            }
```
else? the original code propagates the error.

I see if (!result) below, it's somewhat odd to have an if/else split into two ifs. Suggest to add a comment that explains the failure case here.

piodul replied:
Added comments to this function which explain that the failure is handled later (in the `if (!result) {}` block).

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -408,11 +408,11 @@ select_statement::do_execute(query_processor& qp,
                         options.get_cql_serialization_format(), *_group_by_cell_indices), std::move(p),
                 [this, page_size, now, timeout](auto& builder, std::unique_ptr<service::pager::query_pager>& p) {
-                    return do_until([&p] {return p->is_exhausted();},
+                    return utils::result_do_until([&p] {return p->is_exhausted();},
```
First user of result_do_until. The do_with() indicates a coroutine here (regardless of resultification). It's also the less hot path of aggregate || nonpaged_filtering.

I suggest to push through a patch that converts this code to a coroutine (in a new helper function to not pollute the caller). It can be done in a separate pull request if you like. Then the resultification will be simpler.

Due to all the continuations it will be faster, too, but in a less-important path.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -671,7 +670,7 @@ indexed_table_select_statement::do_execute_base_query(
         auto &previous_result_size = query_state.previous_result_size;
         auto &next_iteration_size = query_state.next_iteration_size;
-        return repeat([this, is_paged, &previous_result_size, &next_iteration_size, &keys, &key_it, &merger, &qp, &state, &options, cmd, timeout]() {
+        return utils::result_repeat([this, is_paged, &previous_result_size, &next_iteration_size, &keys, &key_it, &merger, &qp, &state, &options, cmd, timeout]() {
```
here too

avikivity commented:
Looks good. I think all the sequential loops are better converted to coroutines rather than wrapped in ever-more-complicated loop wrappers, but I may have missed some.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -3813,18 +3813,21 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
 
             // Waited on indirectly.
-            (void)digest_resolver->done().then(utils::result_into_future<result<>>).then([exec, digest_resolver, timeout, background_repair_check] () mutable {
+            (void)utils::result_futurize_try([&] {
```
Do we need to use result_futurize_try to ignore all exceptions? Does `then_wrapped([]{})` will do it?

piodul replied:
I think this will work:

```c++
.then_wrapped([] (future<result<>>&& f) {
    f.ignore_ready_future();
});
```

piodul replied:
Done

piodul commented:
Changelog v1 -> v2:

- Changed the `exception_container` impl to use `std::shared_ptr` internally
- Added "(void) comments" explaining why we are ignoring results in some cases
- Added comments in `query_singular` to explain how failed results are handled
- Renamed `compatible_result` -> `rebind_result` and `ResultCompatible` -> `ResultRebindableTo`
- Got rid of some `result_try` and `result_futurize_try` uses in cases where it was simpler to handle them in some other way

Branch with previous version: [link](https://github.com/piodul/scylla/tree/read-coordinator-timeouts-returned-as-value-v1)

I'll push a rebased version soon and will update the benchmarks after this.

piodul commented:
Changelog v2 -> v2.1:

- Rebased
- Updated benchmarks. The regression seems to have decreased from 500 insns/op to just 200. I think this is a result of switching to using std::shared_ptr - the results are now smaller and moving them is cheaper - and simplifying the error handling logic.

Previous version: [link](https://github.com/piodul/scylla/tree/read-coordinator-timeouts-returned-as-value-v2)

avikivity commented on utils/result.hh:
```diff
@@ -44,4 +44,16 @@ concept ExceptionContainerResult = bo::is_basic_result<R>::value && ExceptionCon
 concept ExceptionContainerResultFuture = seastar::is_future<F>::value && ExceptionContainerResult<typename F::value_type>;
 
+template<typename L, typename R>
+concept ResultRebindableTo =
```
Patch changelog needs updating.

piodul replied:
Updated

avikivity commented on exceptions/exceptions.hh:
```diff
@@ -348,7 +348,9 @@ public:
-    mutation_write_timeout_exception
+    mutation_write_timeout_exception,
+    read_timeout_exception,
+    read_failure_exception
```
For the future, we can consider creating an std::error_category subclass for scylla errors so we can drop the variant (and instead use switches on the error code). But let's delay that until we complete the conversions, it's less important than converting the full stack.

piodul replied:
Do error codes support additional data apart from the error code itself? For example, `mutation_write_timeout_exception` has multiple fields which includes keyspace name, table name, consistency level, etc. - basically all information needed for the CQL layer to send a WRITE_TIMEOUT response to the client ([link](https://github.com/scylladb/scylla/blob/10880fb0a720e3dcddcb301f8f1c624923e89a03/exceptions/exceptions.hh#L152) to the definition in the code). If they don't, then I'm not sure how we could propagate this information.

Perhaps the situation is better on the replica side... If so, then maybe we could use error codes there.

avikivity replied:
We can have class scylla_error : public std::system_error { extra fields } to store stuff. Or even class scylla_error : public std::exception { enum code, extra fields } and drop the error_category idea, which is only useful if we mix multiple categories (like syscall / dns / whatever error categories, which we don't).

piodul commented on service/pager/query_pager.hh:
```diff
@@ -17,6 +17,8 @@
 #include "cql3/selection/selection.hh"
 #include "service/query_state.hh"
+#include "utils/result.hh"
+#include "exceptions/exceptions.hh"
```
Squashed this commit into "query_pager: resultify relevant methods".

piodul commented:
> > Like for writes, there are two tables: `tbl_noop` and `tbl_fail`. I simulated 10k ops/s read workload and measured the Load metric.
> 
> The load metric isn't so good since Seastar will spin a while if it has nothing to do. Better to measure the statement scheduling group utilization.

Here is the "Time used by instance - statement" metric:

![Zrzut ekranu z 2022-02-22 16-42-35](https://user-images.githubusercontent.com/4950936/155167296-26eb938f-b3d4-4596-9e5c-388c25cc4149.png)

The four hills correspond to, from left to right:

|Version|Table name|Utilization (ms)|
|-|-|-|
|Before PR|tbl_noop|120|
|Before PR|tbl_fail|650|
|After PR|tbl_noop|120|
|After PR|tbl_fail|140|

psarna commented:
> > I would like to state for full disclosure that there is still a small amount of C++ exceptions left that will be thrown on the coordinator node, because the RPC module propagates exceptions from replicas and timeouts by throwing. Fixing this would require extending seastar's API so that it is possible to handle such failures without throwing; alternatively, we could extend concrete RPCs so that they also return an error code etc.
> 
> I think we should add new verbs that return a variant<the_real_result, some_exception_type>. We should do this after we agree on a new unified exception class (if we do) to avoid churn. This makes it clear how errors are serialized (they're just values). Currently rpc type-erases exceptions.

Yes please. Ref: https://github.com/scylladb/seastar/issues/835 , it would be nice to be able to let drivers know what happened at the replicas (.e.g whether one is overloaded)

gleb-cloudius commented:
> 
> I think we should add new verbs that return a variant<the_real_result, some_exception_type>. We should do this after we agree on a new unified exception class (if we do) to avoid churn. This makes it clear how errors are serialized (they're just values). Currently rpc type-erases exceptions.
> 
This is not enough for locally generated timeout exceptions. They are generated by the RPC itself.
